### ISSUE TITLE
DRA: kubelet: support permanent and transient errors

### ISSUE THREAD

**Issue by pohly** [kubernetes/kubernetes#125542]
### What would you like to be added?

Some checks made by the kubelet and/or a DRA driver might determine that a problem is permanent. In that case it makes no sense to keep retrying, which is what the current code does.

We need:
- a way to indicate permanent errors in the gRPC interface
- support in the kubelet for marking the pod as permanently failed

### Why is this needed?

- surface permanent errors more obviously
- avoid wasting resources

**Comment by ffromani** [comment]
/triage accepted
/priority important-soon

because reference to 1.31 work

**Comment by kannon92** [comment]
We had a discussion about this with https://github.com/kubernetes/kubernetes/issues/122300. I think the field to mark a pod as failed would be very useful.

What kind of failures were you thinking here? We have an interest in marking a pod as failed for image pulls and maybe lifecycle. But not sure how DRA fits into that.

**Comment by thockin** [comment]
Perhaps a device is tmporarily unavailable (needs a reflash or something) or the driver which implements that device is not responding (e.g. restarting)?

This came up becuase there was a discussion about:

* User requests a flavor of device
* Scheduler assigns to a node
* Kubelet can't reach the device's driver - does it fail admission or admit?  Today it has to admit
* Driver comes back and says "that device is permanently unavailable"

**Comment by bouaouda-achraf** [comment]
Is someone already working on this ? 

/ assign ?

**Comment by pohly** [comment]
No-one is currently working on this, so feel free to give it a go. But beware that some parts, in particular how kubelet should handle permanent failures, may not have an obvious solution. So some exploration will be needed.

**Comment by bouaouda-achraf** [comment]
/assign 

Thank you for your feedback 🙏

**Comment by johnbelamaric** [comment]
cc @lauralorenz

**Comment by pohly** [comment]
@bouaouda-achraf: do you still intend to work on this?

**Comment by bouaouda-achraf** [comment]
Hey @pohly 
I’m currently swamped with other tasks and won’t be able to give this the attention it needs. I’m going to unassign myself so someone else can take over. Sorry for the inconvenience!

**Comment by nojnhuh** [comment]
I can take a look!

/assign

**Comment by nojnhuh** [comment]
I'm catching up on some of the prior investigation related to this: https://docs.google.com/document/d/1-wJhiNy84w7tzFdo9HqwTu5DrVSuXFLGTUv8FBiRAAc/edit?usp=sharing

If the main goal in this issue is to reduce the amount of work kubelet needs to do on terminally failed pods, we can probably punt on trying to do any of the unbinding/rebinding described by that doc and focus only on how the kubelet will propagate some "failed" status such that it knows not to try to reconcile pods in that state.

Would setting the Pod's `status.phase` to `Failed` achieve that? I see the doc mentions reasons for not doing that, but that's what [admission handlers already do](https://github.com/kubernetes/kubernetes/blob/b1ce2a61b532f0afb14b72cf2a63d85f9a3e5a7d/pkg/kubelet/kubelet.go#L2383-L2387). Then to mitigate https://github.com/kubernetes/kubernetes/issues/72593, how reasonable would it be it to rely on drivers to taint or remove devices which are permanently unavailable such that new pods don't keep falling over the same way?

**Comment by nojnhuh** [comment]
Discussed at the SIG-Node meeting today. I don't remember any overt objection to setting `status.phase=Failed` on pods when a DRA driver returns a permanent error, so I'll draft a KEP proposing that.

**Comment by nojnhuh** [comment]
> Discussed at the SIG-Node meeting today. I don't remember any overt objection to setting `status.phase=Failed` on pods when a DRA driver returns a permanent error, so I'll draft a KEP proposing that.

cc @SergeyKanzhelev @ffromani @dchen1107 @lauralorenz (in case I may not have fully followed the discussion)

**Comment by bart0sh** [comment]
@nojnhuh 
>  I don't remember any overt objection to setting status.phase=Failed on pods when a DRA driver returns a permanent error

If this means terminating the pod without unpreparing resources, it was objected here: https://github.com/kubernetes/kubernetes/issues/129402#issuecomment-2578589679

**Comment by pohly** [comment]
When some NodePrepareResources call succeeded and another fails with a permanent error, then we must ensure that NodeUnprepareResources is called to match the successful calls before marking the pod as `status.phase=Failed`.

NodeUnprepareResources could return a final error, but it's not really different from returning no error: the kubelet then can retry (defeating the purposed of the final error) or stop calling the driver (as if it had returned no error). Perhaps it would be best to clearly document that final errors only have that meaning for NodePrepareResources?

**Comment by nojnhuh** [comment]
Opened a KEP issue focused on only the "driver returns known permanent error from NodePrepareResources" case. Happy to adjust scope if that's too narrow for its own KEP: https://github.com/kubernetes/enhancements/issues/5322
